### DISCUSSION TITLE
jj restore --from / --to options

### DISCUSSION THREAD

**Discussion by avamsi** [jj-vcs/jj#1929] [General]
jj move for example, has pretty obvious --from / --to options but I can't seem to build a similar intuition for restore (or even diffedit and instead always need to check the help page beforehand).

I think part of the confusion comes from being able to read --from / --to either way -- i.e., "restore these paths _from_ the way they're in $revision" or "restore these paths _to_ the way they're in $revision" (if anything, the --to version sounds more natural to me and if there's a difference between these two, it's not obvious to me as a non-native English speaker).

Also, editing the --to revision (which is the current behavior) doesn't seem very natural to me (it's called --to in move because you're moving changes _to_ it, with restore, something like -r seems like a better fit).

tl;dr: thoughts on `jj restore [-r=@] [--to=r-]` (current --to renamed to -r, current --from renamed to --to and --to allowing revsets)? Slightly related: #1298.

**Comment by martinvonz** [comment]
I think `-r` is much less clearly a source or destination than `--from`. Would `jj restore --from X --into Y` be clear?

**Reply by avamsi** [reply]
> I think `-r` is much less clearly a source or destination than `--from`.

While I don't necessarily disagree with this, I do think of restore (like diff) as more of a "single revision" operation -- unlike move for example, which is "naturally" more of a two revision operation. So, thinking in terms of source / destination might already be introducing some bias. (Now, I know technically all these operations involve two revisions, maybe it's the defaulting to parent revisions that is fundamental to diffs that makes me think of it as more of a single revision thing, I'm not sure -- sorry if this doesn't make much sense! 😅).

> Would `jj restore --from X --into Y` be clear?

That would definitely be an improvement in isolation but I also think there's value in aligning the signatures of diff and restore (in this case --into diverging away from --to). Can we just ~drop~ replace --to with -r (or make -r/--revision and --to aliases) in all these operations? `jj restore -r X` would revert whatever `jj diff -r X` shows (`jj diffedit -r X` does the same thing but interactively).

cc:@ilyagr too.

**Reply by ilyagr** [reply]
> Can we just drop --to (or make -r/--revision and --to aliases) in all these operations?

I would like all of these operations to work without touching the working copy commit. For example, I might want to compare or even edit other commits while I run some long compile/test operation on the working copy. This makes something like `--to` necessary, I think.

> I also think there's value in aligning the signatures of diff and restore

Me too, but as the discussion about the first version of #1298 and your original question in this discussion show, it's tricky to cover all the usecases *and* be consistent *and* not be confusing. When  these principles conflict, I think consitency is the easiest one to sacrifice.

That's not to say that consistency is not important. Hopefully we'll come up with a better solution through discussions like these.

> > Would `jj restore --from X --into Y` be clear?

I also think this might be an improvement. Here's a proposal. I think it's imperfect, but it's the best I have at the moment:

* Rename `--to` into `--into` for `jj restore` and `jj diffedit`.
* Make `--into` an alias to `--to` for `jj diff`.
* If #1298 goes in, we could also make `--rp` and `--revision-from-parents` aliases to `-r` for `diff` and `diffedit`.

This doesn't solve all the problems with `-r` from @avamsi's original discussion post, but I think the remaining problems are made a little better by the fact that trying to use `-r` with `--to` or `--from` results in a informative error.

If `-r` is still confusing, we could force the user to type `--rp` for `diff` and `diffedit` instead. If the user says to themselves "revisions from parents" as they type, it should not be as confusing.

**Reply by avamsi** [reply]
> I would like all of these operations to work without touching the working copy commit. For example, I might want to compare or even edit other commits while I run some long compile/test operation on the working copy. This makes something like `--to` necessary, I think.

I also like this behavior but it's not obvious to me why this makes --to necessary, would you mind expanding on that?

**Reply by ilyagr** [reply]
If I use `jj diff --from somewhere` to diff with the working copy commit, I'd also like to have a corresponding operation that works with `another_commit` instead of the working copy. Currently, that's what `jj diff --from somewhere --to another_commit` does.

Analogously for `diffedit`: if there is `diffedit --from somewhere` to edit the working copy, I want to be able to edit another commit in the same way.

**Reply by avamsi** [reply]
Sorry for the confusion, I said drop --to, but what I meant to say was replace --to with -r, so you could still pass `-r another_commit` (of course, diff and diffedit already support -r today but we don't allow mixing --from and -r).

The signature I was proposing in the first comment becomes `jj restore [-r=@] [--from=r-]` (current --to renamed to -r, current --from stays the same but allows revsets). diff and diffedit could also have the same (primary) signature.

**Reply by ilyagr** [reply]
Ah, I see, thanks! It's an interesting idea -- have `-r` act as `--to` does today when `--from` is present.

**Reply by ilyagr** [reply]
For `jj diff` and `jj diffedit`, this sounds fine to me at first glance. I don't use `jj diff --to` or `jj diffedit --to` without `--from` very often.

I think I have used `jj restore --to` without `--from` a few times, but perhaps having to add `--from @` wouldn't be a great burden. 

One problem this has is that the user is too likely to use `jj restore -r somewhere` when they mean `jj restore --from somewhere`, which seems like a potentially dangerous and confusing mistake. I guess we could force the user to do `jj restore --rp somewhere` when there's no `--from`. I'm not sure whether this makes things too confusing, especially if our original goal was to simplify `jj restore`.

**Reply by martinvonz** [reply]
I couldn't guess whether `jj restore -r X` means to restore from or to X, so I really don't like using `-r` for that reason.  I don't personally find the current `jj restore --from/--to` confusing at all, but I understand the argument about interpreting `--to X` as "restore to the state in X". I agree that it would be nice if we could make `jj diff` and `jj restore` have matching arguments. How about using `--source/--destination` instead? One advantage of that is that we the short options make sense (we already use them in `rebase`).

By the way - and this is more related to #1298 - we have also talked about using `-c` when we do something with a revision compared to its parents. Maybe we should also make that change, which I think would affect at least `jj diff -r` and the new `jj restore --restore--from-parents`, and maybe `jj rebase -r`.

**Reply by ilyagr** [reply]
> How about using --source/--destination instead? One advantage of that is that we the short options make sense (we already use them in rebase).

That sounds fine to me. The parallel with rebase is actually a little problematic for me, since these options mean something different there, but I don't think it's a serious problem.

> we have also talked about using `-c` when we do something with a revision compared to its parents

That's an interesting idea, I don't remember having discussed it myself. I like it, I think, especially for `diff` and `diffedit`.

I'm not 100% sure this would solve the `restore` issue entirely, but it might. To make it clear enough that people will not use `jj restore -c` when they mean `--from`, we could make the long version of `-c` to be `--changes-in` rather than just `--change`. I think "restoring the changes in REVISION" is OK English, though it bothers me a little bit that this phrase seems to match the functionality `jj revert` at least as well as what it's intended to mean.

**Reply by avamsi** [reply]
What would the final signature of restore be with -c/--changes-in? `jj restore [-c=@] [--from=c-]`? Or would there still be a --to? This change from -r to -c might apply to backout, diffedit, split, squash (and unsquash) as well?

I think I like -c better than --source / --destination, but remembering / guessing which commands take -c vs -r might be somewhat of an issue (will try this when using jj commands to see if my first guess holds up). Also, I would think some (Google internal) commands would want to use -c as CL(s), is that overlap not a concern?

**Reply by avamsi** [reply]
I was just going through https://github.com/martinvonz/jj/pull/1298#discussion_r1112747106 and it's funny how it's almost exactly the same discussion 😅 (@yuja seems to like `jj reset [-r=@] [--from=r-]`, @martinvonz already enumerated a bunch of commands for -r to -c) -- I actually like overwrite from that discussion, it doesn't come with any user expectations that -r would probably be okay to use but OTOH, it's "unnecessarily" diverges from restore / revert terminology.

**Reply by ilyagr** [reply]
> I think I like -c better than --source / --destination

I think `-c` would go together with either `--to/--from` or `--source/--destination`, so it wouldn't help in the way you wanted.

I suppose it might also address @martinvonz 's concern: `jj restore --changes-in REV` seems to only have one possible meaning. So, maybe having just `--changes-in` and `--from` (with `--changes-in` acting like `--to` when `--from` is specified) is still an option. WDYT? 

I'd be OK with it assuming that users would remember that `-c` means "changes in" rather than "change". I'm not certain about that.

>  I actually like `overwrite` from that discussion

Overwrite might be a decent name if we go with `--changes-in` and `--from` being the only arguments. With `--to` or `--destination`, it sounds awkward.

> I was just going through https://github.com/martinvonz/jj/pull/1298#discussion_r1112747106 and it's funny how it's almost exactly the same discussion 😅
 
Yes, I guess I forgot much of it by now; I should have reread it. But hey, I rediscovered `--changes-in` :).  Also, one advantage of forgetting (in a glass half-full way) is that it's then easier for your understanding to change over time. Not that it seemed to change very much in this case.

> remembering / guessing which commands take -c vs -r might be somewhat of an issue

Yes, I'm more worried about this after rereading the old discussion. `squash` seems like an edge case, for example: I could think of `squash` as either operating on revisions or on changes in a revision.

It could still work for `restore` specifically, instead of `--revisions-from-parents`, if we like.

**Reply by avamsi** [reply]
I agree with pretty much everything you've said, although I think I'm more in favor of dropping --to. And yeah, I like overwrite specifically in the context of using it with -r / -c (and --from) -- "overwrite this revision from this other revision" reads okay, I think.
